src/lib/core/fiber.c

```diff
@@ -893,14 +892,96 @@ fiber_recycle(struct fiber *fiber)
 	fiber->storage.lua.fid_ref = FIBER_LUA_NOREF;
 	unregister_fid(fiber);
 	fiber->fid = 0;
+	/* Set before free to disable truncation system area check. */
+	fiber->gc_initial_size = 0;
 	region_free(&fiber->gc);
+#ifdef ENABLE_BACKTRACE
+	fiber->parent_bt = NULL;
+	fiber->first_alloc_bt = NULL;
+	region_set_callbacks(&fiber->gc, NULL, NULL, NULL);
+#endif /* ENABLE_BACKTRACE */
 	if (!has_custom_stack) {
 		rlist_move_entry(&cord()->dead, fiber, link);
 	} else {
 		cord_add_garbage(cord(), fiber);
 	}
 }
 
+#ifdef ENABLE_BACKTRACE
+/**
+ * Called on allocation on region gc. Saves allocation caller backtrace
+ * to report it later if region gc leak is found.
+ */
+static void
+fiber_on_gc_alloc(struct region *region,
+		  size_t size,
+		  void *opaque)
+{
+	(void)region;
+	(void)size;
+	struct fiber *fiber = opaque;
+
+	assert(region_used(&fiber->gc) >= fiber->gc_initial_size);
+	if (region_used(&fiber->gc) == fiber->gc_initial_size) {
+		/* 1 is to skip not interesting frames. */
+		backtrace_collect(fiber->first_alloc_bt, NULL, 1);
+		assert(fiber->first_alloc_bt->frame_count > 0);
+	}
+}
+
+/**
+ * Called on region gc truncation. Resets previously saved backtrace so
+ * it won't be falsely reported in case of programmatic mistake.
+ */
+static void
+fiber_on_gc_truncate(struct region *region,
+		     size_t used,
+		     void *opaque)
+{
+	(void)region;
+	struct fiber *fiber = opaque;
+
+	assert(used >= fiber->gc_initial_size);
+	if (used == fiber->gc_initial_size)
+		fiber->first_alloc_bt->frame_count = 0;
```

Comment:
Why not reset `first_alloc_bt` to null here?

Reply:
Yeah, it is a footprint of the zero version where `first_alloc_bt` was part of fiber structure. Fixed.

Reply:
Done like that so that we don't have to reallocate the backtrace object on each first allocation.